File: leetcode_java/java/com/goinhn/leetcode/_1_100/Sol35.java
package com.goinhn.leetcode._1_100;

/**
 * 35.搜索插入位置
 *
 * @author goinhn
 * @date 2021-11-06
 **/
public class Sol35 {

    /**
     * 先写出二分查找，再进行改造
     * <p>
     * time:O(logn)
     * space:O(1)
     */
    private static class Solution1 {
        public int searchInsert(int[] nums, int target) {
            int left = 0;
            int right = nums.length - 1;
            int result = nums.length;
            while (left <= right) {
                int mid = (left + right) / 2;
                if (target < nums[mid]) {
                    right = mid;
                    result = mid;
                } else if (target > nums[mid]) {
                    left = mid + 1;
                    result = mid + 1;
                } else {
                    return mid;
                }
            }

            return result;
        }
    }

    /**
     *
     */
    private static class Solution2 {
        public int searchInsert(int[] nums, int target) {
            int left = 0;
            int right = nums.length - 1;
            int result = nums.length;
            while (left <= right) {
                int mid = (left + right) / 2;
                if (target <= nums[mid]) {
                    right = mid;
                    result = mid;
                } else {
                    left = mid + 1;
                    result = left;
                }
            }

            return result;
        }
    }
}

File: leetcode_go/question/_1_100/_35/sol_35.go
// Package _35 搜索插入位置
// @author goinhn
// @data 2022/8/31
package _35

// 根据二分进行改造
// time:O(logn)
// space:O(1)
func searchInsert1(nums []int, target int) int {
	left := 0
	right := len(nums) - 1
	res := len(nums)
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] < target {
			left = mid + 1
			res = left
		} else if nums[mid] > target {
			right = mid - 1
			res = mid
		} else {
			return mid
		}
	}

	return res
}

// 根据二分进行改造
// time:O(logn)
// space:O(1)
func searchInsert2(nums []int, target int) int {
	left := 0
	right := len(nums) - 1
	res := len(nums)
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
			res = mid
		}
	}

	return res
}
